### PR TITLE
Add past state support in Attention Op for GPT-2

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**: 

Add a custom op GptAttention for Attention with past  state input. 

The past state is the hidden state for K and V with shape (2, B, N, S', H), where S' is the sequence length for past state. The output of present state will be (2, B, N, S*, H) where S* = S' + S, and S is the sequence length for current input_ids.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Current Attention does not support past state input and output. GPT2 model need past state to speed up decoding.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/fusion_embedlayer.py [hunk 1]
```diff
@@ -38,9 +38,10 @@ class FusionEmbedLayerNoMask(Fusion):
     """
     def __init__(self,
                  model: OnnxModel,
-                 name: str = "EmbedLayerNormalization(no mask)",
-                 search_op_types="SkipLayerNormalization"):
-        super().__init__(model, name, search_op_types)
+                 name: str = "EmbedLayerNormalization",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_embedlayer.py:41]: name => fused_op_type

File: onnxruntime/python/tools/transformers/fusion_gpt_attention.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                                          inputs=[input, gemm.input[1], gemm.input[2], mask_index, past],
+                                          outputs=[attention_node_name + "_output", present],
+                                          name=attention_node_name)
+        attention_node.domain = "com.microsoft"
+        attention_node.attribute.extend(
+            [helper.make_attribute("num_heads", self.num_heads),
+             helper.make_attribute("unidirectional", 1)])
+
+        matmul_node = helper.make_node('MatMul',
+                                            inputs=[attention_node_name + "_output", gemm_qkv.input[1]],
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_gpt_attention.py:37]: format

File: onnxruntime/python/tools/transformers/optimizer.py [hunk 3]
```diff
@@ -90,8 +91,21 @@ def optimize_by_onnxruntime(onnx_model_path, use_gpu=False, optimized_model_path
     logger.info("Save optimized model by onnxruntime to {}".format(optimized_model_path))
     return optimized_model_path
 
+def get_fusion_statistics(optimized_model_path:str) -> Dict[str,int]:
+    """
+    Get counter of fused operators in optimized model.
+
+    Args:
+        optimized_model_path (str): th path of onnx model.
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/optimizer.py:99]: the

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 4]
```diff
@@ -309,10 +309,46 @@ mask_index shall not be provided.)DOC";
       .Input(1, "weight", "2D input tensor with shape (hidden_size, 3 * hidden_size)", "T")
       .Input(2, "bias", "1D input tensor with shape (3 * hidden_size)", "T")
       .Input(3, "mask_index", "Attention mask index with shape (batch_size).", "M", OpSchema::Optional)
-      .Output(0, "output", "3D output tensor with shape (batch_size, sequence_length, hidden_size)", "T")
+      .Input(4, "past", "past state for key and value with shape (2, batch_size, num_heads, past_sequence_length, head_size).", "T", OpSchema::Optional)
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:312]: >2 [](start = 66, length = 1)

How about split the past state for k and v

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 5]
```diff
@@ -309,10 +309,46 @@ mask_index shall not be provided.)DOC";
...
       .TypeConstraint("T", {"tensor(float)", "tensor(float16)"}, "Constrain input and output types to float tensors.")
       .TypeConstraint("M", {"tensor(int32)"}, "Constrain mask index to integer types")
-      .TypeAndShapeInferenceFunction(ONNX_NAMESPACE::propagateShapeAndTypeFromFirstInput);
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+        if (hasInputShape(ctx, 0)) {
+          propagateShapeFromInputToOutput(ctx, 0, 0);
+
+          if (ctx.getNumOutputs() > 1) {
+            propagateElemTypeFromInputToOutput(ctx, 0, 1);
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:323]: the type of output[1] should be inferred even input[0] has no shape info. this logic should be moved out of the scope of "if (hasInputShape(ctx, 0)) {"

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 4 above)
Comment (tianleiwu) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:312]: The export onnx model has such format

File: onnxruntime/contrib_ops/cpu/bert/attention.cc [hunk 6]
```diff
@@ -91,9 +94,70 @@ Status AttentionBase::CheckInputs(const Tensor* input,
...
+    present_dims.push_back(static_cast<int64_t>(past_sequence_length + sequence_length));
+    present_dims.push_back(past_dims[4]);
+  } else {
+    present_dims.push_back(static_cast<int64_t>(2));
+    present_dims.push_back(static_cast<int64_t>(batch_size));
+    present_dims.push_back(static_cast<int64_t>(num_heads_));
+    present_dims.push_back(static_cast<int64_t>(sequence_length));
+    present_dims.push_back(static_cast<int64_t>(head_size));
+  }
+
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/bert/attention.cc:151]: looks like we can make the code more concise with initializer_list:  
std::vector<int64_t> present_dims{2, batch_size, num_heads_, sequence_length, head_size}; 
  if (nullptr != past) {
    const auto past_dims = past->Shape().GetDims();
    past_sequence_length = static_cast<int>(past_dims[3]);
    present_dims[3] += past_sequence_length; 
}

**Pull request merged**